### ISSUE TITLE
Analyze options for batching inputs & streaming results for LM inference

### ISSUE THREAD

**Issue by andreaskoepf** [LAION-AI/Open-Assistant#655]
We are currently planning a system for hosting our trained assistant-models to offer the first public Open-Assistant web chat-interface.

Before we start developing our full-custom hosting system we first want to evaluate existing options and see if they already could fit our needs or at least be used in the beginning to quickly get going.

One obvious candidate here is hosting a [conversational task](https://huggingface.co/tasks/conversational) via the HF inference API (see parameters here: [detailed_parameters](https://huggingface.co/docs/api-inference/detailed_parameters#conversational-task)). This allows at least some form of [automated batching and handling of parallel requests](https://huggingface.co/docs/api-inference/parallelism). A requirement for the chat-interface is that we need to show model-outputs 'near-realtime' (streaming) to the user (max delay of 1-2s before first words are displayed). The HF API offers a `max_time` and `max_length` parameters (e.g. similar to like [MaxTimeCriteria](https://huggingface.co/transformers/v4.6.0/internal/generation_utils.html#transformers.MaxTimeCriteria)). It is not clear if the parallel/batching options + max_time or other stopping criteria could be realistically used for "incremental" sampling (inference resumption) and what the overhead would be (practical or not?).

More general:

The main challenges of the LM hosting and sampling environment are (auto-)scaling and efficient (GPU) resource utilization, e.g. we have not yet a clear idea about how many concurrent sampling requests could be served with which type of GPUs. There are two main use cases for us: a) We would like to host our assistant models in the cloud and offer it for everyone via a web-chat-interface (+potentially API) .. and b) self-hosting on private machines of users.

Scenario a) has much higher requirements, e.g. hundreds of sampling requests potentially running in parallel on different machines... 

Central questions: How can incoming requests of users be batched and dispatched by the inference machines (queues?) and how do we send streaming results back to the clients (e.g. routing between web-endpoints and inference-machines, websocket connections to browser clients?).

Suggested action items:

- play with the HF API
- gather architecture ideas for a scalable distributed batching & sampling system
- if possible create a prototype of a scalable inference system

(A meeting of the backend team is scheduled for 2023-01-12 @ 17:30 UTC, if you are interested please ping me on the OA discord server .. see main repo README)

**Comment by Nil-Andreu** [comment]
Okay, the first idea (this is like a draft) that comes in my mind is the following:
1. Service that gets the message via websocket and creates a new task in a task-queue (like Redis).
2. There is a worker that will be reading from this task-queue, which batch the tasks, and once it obtains a batch it makes a call to the ML API (this case we can auto-scale the workers if we need more speed).
3. With the result back, maybe we could use a message (RabbitMQ, KafKa) broker to read back which is the response for that particular message, and send it back to the frontend.

I will try to get around this issue, and if anyone wants to add more the feedback is appreciated!

**Comment by olliestanley** [comment]
Quick overview of what I have looked into so far for hosting the model itself:

**HF Inference API** will not load large models. Pythia 13B [which Rallio's model is derived from] times out when loading. 6.7B, 2.7B, 1.3B, 350m also time out when loading. 125m loads and runs successfully.

**HF Inference Endpoints** is the production counterpart which requires a lot more setup and has no free access that I can see, but means you can pick specific instance types and therefore would be able to load large models. Behind the scenes it looks like this is basically a wrapper around SageMaker Endpoints so it may be worth comparing this approach against using SageMaker directly.

It is possible we could make a response streaming approach work with HF Inference Endpoints or SageMaker Endpoints but it would have to be hacky (such as repeated requests each generating a few tokens which are then appended to the next request) as these services are set up to assume a single "return value" from each request to the model.

**Comment by alramalho** [comment]
- Was reading the [blog post](url) by AWS Sagemaker, and it looks like it would be a viable option, if we have the budget. `EleutherAI/gpt-j-6B`was setup along with DeepSpeed Tensor paralellism to allow for inference of under 0.1s in two GPUs of the `ml.p3.16xlarge` instance, which has `64 vCPUs`, `488 GiB` of Memory, has `8 x V100 GPUS`  at `16GB / GPU` and runs at `$28,152 per hour`, no batch processing or streaming.
    - [source1](https://pages.awscloud.com/rs/112-TZM-766/images/AL-ML%20for%20Startups%20-%20Select%20the%20Right%20ML%20Instance.pdf), [source2](https://aws.amazon.com/pt/sagemaker/pricing/)
    - there’s extra optimization techniques that are not being used here, such as quantization. This holds potential to reduce the instance size and lower costs
    
- Looks like Cohere also has [some optimization techniques going on](https://txt.cohere.ai/running-large-language-models-in-production-a-look-at-the-inference-framework-tif/), which would suggest fast inference times. I did not find any example in regards to these, and no mention to batch processing or streaming requests was found in their docs

**Comment by OlivierDehaene** [comment]
Just to chime in the discussion: we developped at Hugging Face a specific backend to handle dynamic batching for the LLMs hosted on our platform: [text-generation-inference](https://github.com/huggingface/text-generation-inference).

This backend works by batching incoming requests together, evicting them when they meet their stopping criteria and re-adding new requests to the current batch to mitigate noisy neighbor problems where one request asks for a lot more tokens than the others.

It also handles model sharding/TP. For now, we add TP manually but it is easy to plug DeepSpeed (we had some issues with DS inference in the past so we removed it for now).

We are considering adding token streaming if you guys are interested.

> HF Inference API will not load large models.

By default we do not. However, we have exceptions in place for popular models were we plug more powerful machines to make them available/faster.

**Comment by yk** [comment]
Thanks @OlivierDehaene for the tip. I've seen that and started working on a fork here to include streaming: https://github.com/yk/text-generation-inference

In case you think it's a good direction, maybe we could work towards an eventual PR to HF's repo.

**Comment by OlivierDehaene** [comment]
Yes this is very close to what I had in mind! 
If you are ok with this I will continue your work in a fork of your fork tomorrow and merge this asap :)

**Comment by yk** [comment]
fyi I now also added rng seeds and some other stuff

**Comment by OlivierDehaene** [comment]
[This is roughly what I had in mind](https://github.com/huggingface/text-generation-inference/pull/36). There are still some things to iron out.

One question I have is regarding the API. I think the only events we care about while streaming are token events. The user does not really care about the prompt tokens (prefill in my PR) and the generated text is redondant. What do you think?

> I also added rng seeds

Feel free to make a PR anytime :)

**Comment by OlivierDehaene** [comment]
@yk, I'm done with my implem [here](https://github.com/huggingface/text-generation-inference/pull/36).
Does the following SSE event signature cover your usecase?

```rust
struct StreamToken {
    token: (u32, String, f32),
    end: bool,
    finish_reason: Option<String>,
    generated_text: Option<String>,
}
```

**Comment by alando46** [comment]
What are the remaining todo's on this effort? It seems like there was coalescing around using HF's tool for mvp and maybe the needed streaming was added last week (https://github.com/huggingface/text-generation-inference/releases/tag/v0.2.0)? 

Anything else needed to pick up?
cc @yk @olliestanley

**Comment by yk** [comment]
this is largely implemented now in the `/inference` folder
